noir_stdlib/src/meta/expr.nr

```diff
@@ -110,4 +110,305 @@ impl Expr {
+        let index = index.map(f);
+        new_index(object, index)
+    }
+    )
+}
+
+fn map_member_access<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_member_access().map(
+        |expr: (Expr, Quoted)| {
+        let (object, name) = expr;
+        let object = object.map(f);
+        new_member_access(object, name)
+    }
+    )
+}
+
+fn map_method_call<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_method_call().map(
+        |expr: (Expr, Quoted, [UnresolvedType], [Expr])| {
+        let (object, name, generics, arguments) = expr;
+        let object = object.map(f);
+        let arguments = arguments.map(|arg: Expr| arg.map(f));
+        new_method_call(object, name, generics, arguments)
+    }
+    )
+}
+
+fn map_repeated_element_array<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_repeated_element_array().map(
+        |expr: (Expr, Expr)| {
+        let (expr, length) = expr;
+        let expr = expr.map(f);
+        let length = length.map(f);
+        new_repeated_element_array(expr, length)
+    }
+    )
+}
+
+fn map_repeated_element_slice<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_repeated_element_slice().map(
+        |expr: (Expr, Expr)| {
+        let (expr, length) = expr;
+        let expr = expr.map(f);
+        let length = length.map(f);
+        new_repeated_element_slice(expr, length)
+    }
+    )
+}
+
+fn map_slice<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_slice().map(
+        |exprs: [Expr]| {
+        let exprs = map_expressions(exprs, f);
+        new_slice(exprs)
+    }
+    )
+}
+
+fn map_tuple<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_tuple().map(
+        |exprs: [Expr]| {
+        let exprs = map_expressions(exprs, f);
+        new_tuple(exprs)
+    }
+    )
+}
+
+fn map_unary_op<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_unary_op().map(
+        |expr: (UnaryOp, Expr)| {
+        let (op, rhs) = expr;
+        let rhs = rhs.map(f);
+        new_unary_op(op, rhs)
+    }
+    )
+}
+
+fn map_unsafe<Env>(expr: Expr, f: fn[Env](Expr) -> Option<Expr>) -> Option<Expr> {
+    expr.as_unsafe().map(
+        |exprs: [Expr]| {
+        let exprs = exprs.map(|expr: Expr| expr.map(f));
+        new_unsafe(exprs)
+    }
+    )
+}
+
+fn map_expressions<Env>(exprs: [Expr], f: fn[Env](Expr) -> Option<Expr>) -> [Expr] {
+    exprs.map(|expr: Expr| expr.map(f))
+}
+
+fn new_array(exprs: [Expr]) -> Expr {
+    let exprs = join_expressions(exprs, quote { , });
+    quote { [$exprs]}.as_expr().unwrap()
+}
+
+fn new_assign(lhs: Expr, rhs: Expr) -> Expr {
+    quote { $lhs = $rhs }.as_expr().unwrap()
+}
+
+fn new_binary_op(lhs: Expr, op: BinaryOp, rhs: Expr) -> Expr {
+    let op = op.quoted();
+    quote { $lhs $op $rhs }.as_expr().unwrap()
```

Comment:
Probably want to parenthesize this. Otherwise these could combine in odd ways with precedence.